src/ens/storage.c

```diff
@@ -0,0 +1,175 @@
+        printk("Cannot retrieve page information (err %d)\n", rc);
+        return rc;
+    }
+    info_fs.sector_size = info.size;
+    info_fs.sector_count = SEC_COUNT;
+
+    rc = nvs_init(&info_fs, DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL);
+    k_mutex_init(&info_fs_lock);
+    if (rc) {
+        // Error during nvs_init
+        printk("Cannot init NVS (err %d)\n", rc);
+        return rc;
+    }
+
+    // Load the current storage information
+    rc = load_storage_information();
+    if (rc) {
+        printk("Cannot load storage information (err %d)\n", rc);
+        return rc;
+    }
+
+    printk("Currently %d contacts stored!\n", contact_information.count);
+    printk("Space available: %d\n", FLASH_AREA_SIZE(storage));
+
+    // TODO lome: change size to sizeof(contact_struct)
+    rc = ens_fs_init(&ens_fs, FLASH_AREA_ID(ens_storage), 32);
+    if (rc) {
+        printk("Cannot init ens_fs (err %d)\n", rc);
+    }
+    return rc;
+}
+
+int load_contact(record_t* dest, record_sequence_number_t sn) {
+    storage_id_t id = convert_sn_to_storage_id(sn);
+    // int rc = nvs_read(&info_fs, id, dest, sizeof(*dest));
+    int rc = ens_fs_read(&ens_fs, id, dest);
+    if (rc < 0) {
+        return rc;
+    }
+    return 0;
+}
+
+int add_contact(record_t* src) {
+    // Check, if next sn would be at start of page
+    record_sequence_number_t potential_next_sn = sn_increment(get_latest_sequence_number());
+    storage_id_t potential_next_id = convert_sn_to_storage_id(potential_next_sn);
+    if (((potential_next_id * ens_fs.entry_size) % ens_fs.sector_size) == 0) {
+        // If we are at start of a page, we need to erase it first
+        ens_fs_page_erase(&ens_fs, potential_next_id, 1);
+    }
+
+    // Actually increment sn
+    record_sequence_number_t curr_sn = get_next_sequence_number();
+    src->sn = curr_sn;
+    storage_id_t id = convert_sn_to_storage_id(curr_sn);
+
+    return ens_fs_write(&ens_fs, id, src);
+}
+
+int delete_contact(record_sequence_number_t sn) {
+    storage_id_t id = convert_sn_to_storage_id(sn);
+    int rc = ens_fs_delete(&ens_fs, id);
+    if (!rc) {
+        k_mutex_lock(&info_fs_lock, K_FOREVER);
+        if (sn_equal(sn, get_latest_sequence_number())) {
+            contact_information.count--;
+        } else if (sn_equal(sn, get_oldest_sequence_number())) {
+            contact_information.oldest_contact = sn_increment(contact_information.oldest_contact);
+            contact_information.count--;
+        }
+        save_storage_information();
+        k_mutex_unlock(&info_fs_lock);
+    }
+    return rc;
+}
+
+// TODO lome: do we need lock here aswell?
+record_sequence_number_t get_latest_sequence_number() {
+    return sn_mask(contact_information.oldest_contact + contact_information.count);
+}
+
+record_sequence_number_t get_oldest_sequence_number() {
+    return contact_information.oldest_contact;
+}
+
+uint32_t get_num_contacts() {
+    return contact_information.count;
```

Comment:
We might want to distinguish between the real amount of contacts and the amount of used sequence numbers, maybe?